### PR TITLE
[CUB] Dmitrii Artiukhov A02

### PR DESCRIPTION
**Pull request by dmitrii-artuhov**

### CODE REVIEW

File: src/X86.lama [hunk 1]
```diff
@@ -272,6 +272,79 @@ fun suffix (op) {
...
+fun compileAssociativeOp (op, x, y, env, code) {
+  if stackOpnd(x) && stackOpnd(y) then [env.push(y), code <+ Mov (x, eax) <+ Binop (op, eax, y)]
+  else [env.push(y), code <+ Binop (op, x, y)]
+  fi
+}
+
+fun compileDivisionOp (op, x, y, env, code) {
+  -- we are actually doing (y / x) or (y % x) here
+  case op of 
+    "/" -> [env.push(y), code <+ Mov (y, eax) <+ IDiv (x) <+ Mov (eax, y)]
```
Comment (e2e4b6b7) [inline src/X86.lama:284]: Cltd

File: src/X86.lama [hunk 2]
```diff
@@ -272,6 +272,79 @@ fun suffix (op) {
...
+
+fun compileLogicalOp (op, x, y, env, code) {
+  -- if y == 0: env.push(0) else: env.push(1)
+  -- if x == 0: env.push(0) else: env.push(1)
+  -- return env.pop2 -> [binx, biny, env]: compileBinop(op, binx, biny, env, code)
+
+  fun compileConvertionIntToBoolean (pos, env, code) {
+    var dl = "%dl";
+
+    if stackOpnd(pos) then [env.push(pos), code <+ Mov(L (0), edx) <+ Mov (pos, eax) <+ Binop ("cmp", L (0), eax) <+ Set (suffix("!="), dl) <+ Mov (edx, pos)]
```
Comment (e2e4b6b7) [inline src/X86.lama:297]: redundant branch. cmp allows to compare immediate operand with memory location

File: src/X86.lama [hunk 3]
```diff
@@ -272,6 +272,79 @@ fun suffix (op) {
...
+            [binx, biny, env] -> compileAssociativeOp(op, binx, biny, env, code)
+          esac
+      esac
+  esac
+}
+
+fun compileComparisonOp(op, x, y, env, code) {
+  var dl = "%dl";
+
+  if (stackOpnd(x)) then [env.push(y), code <+ Mov (L (0), edx) <+ Mov (x, eax) <+ Binop ("cmp", eax, y) <+ Set (suffix(op), dl) <+ Mov (edx, y)]
```
Comment (e2e4b6b7) [inline src/X86.lama:316]: if x is stack operand and y is in register, the move is redundant

**Review by e2e4b6b7 — CHANGES_REQUESTED** [review]

**Comment by e2e4b6b7** [comment]
Fix this task until the end of Monday or it will considered as failed

File: src/X86.lama (hunk 1 above)
Comment (dmitrii-artuhov) [reply src/X86.lama:284]: done

File: src/X86.lama (hunk 2 above)
Comment (dmitrii-artuhov) [reply src/X86.lama:297]: done

**Pull request closed without merging**